debezium-core/src/main/java/io/debezium/relational/history/KafkaDatabaseHistory.java

```diff
@@ -224,16 +231,33 @@ protected void recoverRecords(Tables schema, DdlParser ddlParser, Consumer<Histo
                     }
                 }
                 if (numRecordsProcessed == 0) {
-                    --remainingEmptyPollResults;
-                    logger.debug("No new records found in the database history; will retry {} more times", remainingEmptyPollResults);
+                    logger.debug("No new records found in the database history; will retry");
                 } else {
-                    remainingEmptyPollResults = this.recoveryAttempts;
                     logger.debug("Processed {} records from database history", numRecordsProcessed);
                 }
+
+                recoveryAttempts++;
             }
+            while (lastProcessedOffset < endOffset - 1);
         }
     }
 
+    private Long getEndOffsetOfDbHistoryTopic(Long previousEndOffset, KafkaConsumer<String, String> historyConsumer) {
+        Map<TopicPartition, Long> offsets = historyConsumer.endOffsets(Collections.singleton(new TopicPartition(topicName, PARTITION)));
+        Long endOffset = offsets.entrySet().iterator().next().getValue();
+
+        // The end offset should never change during recovery; doing this check here just as - a rather weak - attempt
+        // to spot other connectors that share the same history topic accidentally
+        if(previousEndOffset != null && !previousEndOffset.equals(endOffset)) {
+            throw new IllegalStateException("Detected changed end offset of database history topic (previous: "
+                    + previousEndOffset + ", current: " + endOffset
+                    + "). Make sure that the same history topic isn't shared by multiple connector instances."
```

Comment:
We do share the same history topic with multiple connector instances but will be changing this to accommodate the requirement. I'm not sure if this requirement is officially documented.

Reply:
I agree it might be nice to not have this requirement, but we currently do. We would need to add keys to allow distinguishing from multiple connectors, but it's doable.

Reply:
> I'm not sure if this requirement is officially documented.

It's mentioned via the `database.server.name` property (which is used as topic name), where we say it should be unique.

In terms of sharing the topic across connectors, I think it'd add quite some complexity. Not sure whether it's worth it.

Slightly related to this, I noticed a potential issue where we'd use an incorrect schema for some messages when the connector restarts after a crash: [DBZ-488](https://issues.jboss.org/browse/DBZ-488). The reason being that we recover the history until "latest", whereas the connector may resume at an earlier binlog position (at the last committed offset).